### DISCUSSION TITLE
Crash task-local: illegal task-local value binding

### DISCUSSION THREAD

**Discussion by corysullivan** [pointfreeco/swift-dependencies#110] [Q&A] [answered]
I am updating my app to TCA 1.0 and using the latest Xcode beta (Version 15.0 beta 6 (15A5219j)). After resolving all compilation errors, I encountered a crash from the Dependency library. I intend to investigate this further, but I wanted to know if anyone else has experienced this or has any ideas on why it might happen.

<img width="1758" alt="Screenshot 2023-08-10 at 1 26 41 PM" src="https://github.com/pointfreeco/swift-dependencies/assets/1440684/bca73603-bdda-44fd-aa6c-3623681fbf52">

**Comment by stephencelis** [comment, marked as answer]
@corysullivan That error is due to executing task local code inside an invalid context, like within a `withTaskGroup` block but outside an `addTask`. Here's a vanilla test case demonstrating the issue:

<details>
<summary><img width="904" alt="image" src="https://github.com/pointfreeco/swift-dependencies/assets/658/8000b3a7-29b9-4cea-a448-840e0692fa92"></summary>

```swift
func testIllegalTaskLocal() async {
  enum Test {
    @TaskLocal static var isSetting = false
  }

  await withTaskGroup(of: Void.self) { group in
    Test.$isSetting.withValue(true) {
      print(Test.isSetting)
    }
    print(Test.isSetting)
  }
}
```

</details>

If you follow your trace to the effect causing the crash, you're likely in a `withTaskGroup` context. You'll want to avoid accessing dependencies in that scope, and instead move dependency access inside `group.addTask`, or outside the task group entirely.

My only guess as to why you didn't encounter this issue earlier is that you were on a version of Dependencies that didn't access the task local at this time.

**Reply by corysullivan** [reply]
Thanks Stephen, This was helpful, I found the issue, and although I wasn't accessing a dependency directly in the group context, I was sending an action back into the system.
```swift
return .run { send in
  await withTaskGroup(of: Void.self) { group in
    for task in tasks {
      group.addTask {
        ...
      }
    }
    await send(.someAction)
  }
}
```
Moving the action sending outside the group context resolved the issue.
